tokio-util/src/sync/cancellation_token/implementation.rs

```diff
@@ -0,0 +1,341 @@
+//! GENERAL NOTES
+//! -------------
+//!
+//! ** Invariants **
+//!
+//! Those invariants shall be true at any time, and are required to prove correctness
+//! of the CancellationToken.
+//!
+//! 1. A node that has no parents and no handles can no longer be cancelled.
+//!     This is important during both cancellation and refcounting.
+//!
+//! 2. If node B *is* or *was* a child of node A, then node B was created *after* node A.
+//!     This is important for deadlock safety, as it is used for lock order.
+//!     Node B can only become the child of node A in two ways:
+//!         - being created with `child_node()`, in which case it is trivially true that
+//!           node A already existed when node B was created
+//!         - being moved A->C->B to A->B because node C was removed in `decrease_handle_refcount()`.
+//!           In this case the invariant still holds, as B was younger than C, and C was younger
+//!           than A, therefore B is also younger than A.
+//!
+//! 3. If two nodes are both unlocked and node A is the parent of node B, then node B is a child of node A.
+//!     It is important to always restore that invariant before dropping the lock of a node.
+//!
+//! ** Deadlock safety **
+//!
+//! Principles that provide deadlock safety:
+//!     1. We always lock in the order of creation time. We can prove this through invariant #2.
+//!        Specifically, through invariant #2, we know that we always have to lock a parent before its child.
+//!     2. We never lock two siblings simultaneously, because we cannot establish an order.
+//!        There is one exception in `with_locked_node_and_parent()`, which is described in the function itself.
```

Comment:
I would simplify this to just say that we always lock in order of creation, which we can determine through invariant #2. You don't need to talk about siblings here.